src/entry.lisp

```diff
@@ -194,23 +173,96 @@
 
-(defun file-entry-point (filename)
-  (declare (type string filename))
-
-  (with-open-file (file-stream filename :if-does-not-exist :error)
-    (let ((coalton-file (error:make-coalton-file
-                         :stream file-stream
-                         :name filename)))
-      (multiple-value-bind (code env)
-          (entry-point (parser:read-program file-stream coalton-file :mode :file))
-
-        (setf *global-environment* env)
-
-        code))))
-
-(defun debug-file-entry-point (filename)
-  (declare (type string filename))
-
-  (let ((settings:*coalton-skip-update* t)
-        (settings:*emit-type-annotations* nil))
-    (file-entry-point filename)))
+(defmacro with-environment-updates (updates &body body)
+  "Collect environment updates into a vector bound to UPDATES."
+  `(let* ((,updates (make-array 0 :adjustable t :fill-pointer 0))
+          (tc:*update-hook* (lambda (name arg-list)
+                              (vector-push-extend (cons name arg-list) ,updates))))
+     ,@body))
+
+(defun make-environment-updater (update-log)
+  "Produce source form of the contents of an environment UPDATE-LOG (i.e., calls to functions in typechecker/environment)."
+  (let ((updates (remove-duplicates (coerce update-log 'list) :test #'code-update-eql)))
+    `(let ((env *global-environment*))
+       ,@(loop :for (fn . args) :in updates
+               :collect `(setf env (,fn env ,@(mapcar #'util:runtime-quote args))))
+       (setf *global-environment* env))))
+
+(defun compile-coalton-toplevel (program)
+  "Compile PROGRAM and return a single form suitable for direct inclusion by Lisp compiler. For implementing coalton-toplevel macro."
+  (with-environment-updates updates
+    (multiple-value-bind (program env)
+        (entry-point program)
+      (setf *global-environment* env)
+      `(progn
+         ,(make-prologue)
+         ,(make-environment-updater updates)
+         ,program))))
+
+(defun code-update-eql (a b)
+  "Compare environment updates, returning t for set-code updates of the same symbol."
+  (and (eql (first a) 'coalton-impl/typechecker/environment:set-code)
+       (eql (first b) 'coalton-impl/typechecker/environment:set-code)
+       (eql (second a)
+            (second b))))
+
+(defun make-prologue ()
+  "Return source form of an assertion that prevents mixing development and release modes."
+  `(eval-when (:load-toplevel)
+     (unless (eq (settings:coalton-release-p) ,(settings:coalton-release-p))
+       ,(if (settings:coalton-release-p)
+            `(error "~A was compiled in release mode but loaded in development."
+                    ,(or *compile-file-pathname* *load-truename*))
+            `(error "~A was compiled in development mode but loaded in release."
+                    ,(or *compile-file-pathname* *load-truename*))))))
+
+(defun print-form (stream form)
+  "Print a FORM to a STREAM, separated by 2 lines."
+  (let ((*package* (find-package :cl)))
```

Comment:
`*package* is going to get reset to `cl-user` with `with-standard-io-syntax`. I would instead do:

```
(with-standard-io-syntax
  (let ((*package* (find-package "KEYWORD")))
    (prin1 ...) ...))
```